neo/SmartContract/Native/Tokens/Nep5Token.cs

```diff
@@ -0,0 +1,152 @@
+﻿using Neo.Ledger;
+using Neo.VM;
+using System;
+using System.Numerics;
+using VMArray = Neo.VM.Types.Array;
+
+namespace Neo.SmartContract.Native.Tokens
+{
+    public abstract class Nep5Token<TState> : NativeContract
+        where TState : Nep5AccountState, new()
+    {
+        public override ContractPropertyState Properties => ContractPropertyState.HasStorage;
+        public override string[] SupportedStandards { get; } = { "NEP-5", "NEP-10" };
+        public abstract string Name { get; }
+        public abstract string Symbol { get; }
+        public abstract int Decimals { get; }
+        public BigInteger Factor { get; }
+
+        protected const byte Prefix_TotalSupply = 11;
+        protected const byte Prefix_Account = 20;
+
+        protected Nep5Token()
+        {
+            this.Factor = BigInteger.Pow(10, Decimals);
+        }
+
+        protected StorageKey CreateAccountKey(UInt160 account)
+        {
+            return CreateStorageKey(Prefix_Account, account);
+        }
+
+        protected override StackItem Main(ApplicationEngine engine, string operation, VMArray args)
+        {
+            switch (operation)
+            {
+                case "name":
+                    return Name;
+                case "symbol":
+                    return Symbol;
+                case "decimals":
+                    return Decimals;
+                case "totalSupply":
+                    return TotalSupply(engine);
+                case "balanceOf":
+                    return BalanceOf(engine, new UInt160(args[0].GetByteArray()));
+                case "transfer":
+                    return Transfer(engine, new UInt160(args[0].GetByteArray()), new UInt160(args[1].GetByteArray()), args[2].GetBigInteger());
+                default:
+                    return base.Main(engine, operation, args);
+            }
+        }
+
+        internal protected virtual void MintTokens(ApplicationEngine engine, UInt160 account, BigInteger amount)
```

Comment:
`internal protected` looks like an amazing solution... is it visible on UT, right?

Reply:
Yes, it is visible to UT.